discv5/discv5-wire.md

```diff
@@ -49,187 +49,113 @@ request/response and 1s for the handshake.
+    authdata-size = 52
+    request-nonce = uint96    -- nonce of request packet that couldn't be decrypted
+    id-nonce      = uint256   -- random bytes
+    enr-seq       = uint64    -- ENR sequence number of the requesting node
 
-The first encrypted message sent in response to WHOAREYOU contains an authentication
-header completing the handshake. The plain text of the authentication response is.
+![whoareyou packet layout](./img/whoareyou-packet-layout.png)
 
-    auth-response-pt = [version, id-nonce-sig, node-record]
-    version          = 5
-    id-nonce-input   = sha256("discovery-id-nonce" || id-nonce || ephemeral-key)
-    id-nonce-sig     = id_sign(id-nonce-input)
-    static-node-key  = the private key used for node record identity
-    node-record      = record of sender OR [] if enr-seq in WHOAREYOU != current seq
+### Handshake Message Packet (`flag = 2`)
 
-`auth-response-pt` is encrypted with a separate key, the `auth-resp-key`, using an
-all-zero nonce. This is safe because only one message is ever encrypted with this key.
+For handshake message packets, the `authdata` section has variable size since public key
+and signature sizes depend on the ENR identity scheme. For the "v4" identity scheme, we
+assume 64-byte signature size and 33 bytes of (compressed) public key size.
 
-    message-packet   = tag || auth-header || message
-    auth-header      = [auth-tag, id-nonce, auth-scheme-name, ephemeral-pubkey, auth-response]
-    auth-scheme-name = "gcm"
-    auth-response    = aesgcm_encrypt(auth-resp-key, zero-nonce, auth-response-pt, "")
-    zero-nonce       = 12 zero bytes
-    message          = aesgcm_encrypt(initiator-key, auth-tag, message-pt, tag)
-    message-pt       = message-type || message-data
-    auth-tag         = AES-GCM nonce, 12 random bytes unique to message
+`authdata` starts with a fixed-size `authdata-head` component, followed by the ID
+signature, ephemeral public key and optional node record.
 
-All messages following the handshake are encoded as follows:
+The `record` field may be omitted if the `enr-seq` of WHOAREYOU is recent enough, i.e.
+when it matches the current sequence number of the sending node. If `enr-seq` is zero, the
+record must be sent. Node records are encoded and verified as specified in [EIP-778].
 
-    message-packet   = tag || rlp_bytes(auth-tag) || message
-    message          = aesgcm_encrypt(initiator-key, auth-tag, message-pt, tag)
+Please refer to the [handshake section] for more information about the content of the
+handshake packet.
 
-Implementations can distinguish the two `message-packet` encodings by checking whether the
-value at offset 32 after the fixed-size `tag` is an RLP list (`auth-header`) or byte array
-(`auth-tag`).
+    authdata      = authdata-head || id-signature || eph-pubkey || record
+    authdata-size = 15 + sig-size + eph-key-size + len(record)
+    authdata-head = version || nonce || sig-size || eph-key-size
```

Comment:
`authdata-head` should also contain plain `id-nonce` string. Making `id-nonce` validation through signature verification opens a DoS vector. An adversary will be able to create handshake packet once and force remote peers to verify signatures consuming their CPU.

A particular attack scenario outlined by @Nashatyrev: malicious peer sends a couple of packets with a small timeout between them, first one is treated as a random packet and initiates the handshake process, second packet is the handshake packet which forces honest peer to verify signature before it's able to recognise an invalid `id-nonce`. In this attack the same message can be used for both packets, it can be reused between different targets and attack rounds.

Reply:
I don't understand how including `id-nonce` here prevents this.  It seems that you could still issue a bogus handshake response with a bogus signature in this case and that the bogus signature can still be re-used across different target attacks.  It's unclear to me what including the `id-nonce` here does to prevent that.

Reply:
Sure, you can, but if `id-nonce` is not included it makes this attack much more efficient. An attacker will have to wait for the `WHOAREYOU` message to include valid `id-nonce` to be able to pass an initial check and trigger signature verification.

We can't completely eliminate the DoS vector, ratelimiting seems to be the only way to mitigate it.

Reply:
In general, the effectiveness of this attack also depends on the performance of signature verification. In discv4, we have a signature on every packet, and this attack was never considered because signature verification is fast. Sure, it might be possible to reject the bad packet even faster if including the nonce is required, but it will also increase the size of the handshake packet. We need to weigh these factors. I'm personally not to worried about this particular attack
because I think signature verification is fast enough.

Reply:
The only thing the proposed solution seems to fix is that it forces the packet to be sent **after** the WHOAREYOU meaning that in both situations one must still do signature verification so it seems that the speed of signature verification doesn't actually matter.

It seems that all that matters here is whether the increased cost of the attacker having to wait for the WHOAREYOU packet prior to sending the bogus HANDSHAKE packet makes a meaningful difference.  In both cases the recipient will have to perform signature verification.  This leads me to think that this may be better addressed via some form of artificial rate limiting on inbound handshakes.

My opinion is that we should not make the proposed protocol change but I'm open to understanding where my reasoning is off if someone sees something here that I'm missing.

Reply:
> In discv4, we have a signature on every packet, and this attack was never considered because signature verification is fast. 

That is true, signature verification takes ~1ms on my laptop with core i7 CPU.

I have the following concern. Relatively cheap [spoofed VPS](https://www.clientvps.com/service/spoofed-vps) could turn this DoS vector into, effectively, DDoS  vector, i.e. one server becomes able to attack from tons of remote peers triggering tons of signature verifications per unit of time. With `id-nonce` in place an attacker will have to read `WHOAREYOU` message which are sent to IPs that he has no ownership over hence a single machine attacker will be able to trigger sig verification for only one handshake at a time.

> This leads me to think that this may be better addressed via some form of artificial rate limiting on inbound handshakes.

Definitely. Artificial rate limiting is the protection that we want in the first place. And it might also be enough to mitigate that DoS leveraged by spoofing.

> but it will also increase the size of the handshake packet

It's gonna be 16 bytes (with `IV` replacing `id-nonce`) per session. Is this a significant increase for the handshake packet?

Reply:
I think I'm +0 on this having thought about it more.  Requiring the `id-nonce` value to be present in the HANDSHAKE packet seems to provide at least *some* increased cost to executing the attack.

The reason I'm not sold is this.  Even with the change, an attacker can just include a random value.  While a proper implementation can avoid the extra signature verification since the `id-nonce` check will fail to match, the attacker can still force the initial message packet to be parsed, the WHOAREYOU packet to be generated and sent, and the HANDSHAKE packet to be parsed and `id-nonce` checked.  The **only** thing this seems to do is *maybe* save one signature verification which we've already established is inexpensive... 

I do however see that this is a small change, and the cost of including the `id-nonce` in the HANDSHAKE header is indeed only 16 bytes and it does seem to provide *some* improvement.  I'm good with whatever decision is made on this (which I'm :+1: on letting @fjl make the final call).

Reply:
I'm still thinking about it.

Reply:
I have decided to not address it in v5.1. This is something we can change in v5.2.

Reply:
I just realized there is a good way to address this without increasing the handshake packet size even more: just replace the src-id with id-nonce.

Reply:
> I just realized there is a good way to address this without increasing the handshake packet size even more: just replace the src-id with id-nonce. We can definitely consider it for v5.2.

I can file an issue with this particular proposal. So we won't forget to consider it during further update.

Reply:
Yes, please do that!